exe/task.cc

```diff
@@ -137,6 +148,161 @@ static map<string, string> generate_environment_from(lth_jc::JsonContainer &inpu
+// by just calling calculate_sha256 instead.
+/*static*/ size_t write_data(void *ptr, size_t size, size_t nmemb, FILE *stream) {
+    return fwrite(ptr, size, nmemb, stream);
+}
+
+// This method downloads the file at the specified url into the provided path
+// If something goes wrong during the download, a PXP-error should be returned.
+/*static*/ void download_task_file(const fs::path& path, const string& url) {
+    auto curl = curl_easy_init();
+    if (!curl) {
+      // TODO: Throw PXP-error here? Means something went wrong w/ initializing the download.
+      return;
+    }
+  
+    char err_buff[CURL_ERROR_SIZE];
+    auto fp = fopen(path.string().c_str(),"wb");
+    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
+    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_data);
+    curl_easy_setopt(curl, CURLOPT_WRITEDATA, fp);
+    // We can get an error message this way if something goes wrong in the download
+    curl_easy_setopt(curl, CURLOPT_ERRORBUFFER, err_buff);
+    // CONNECTTIMEOUT specifies how long curl should wait for a connection to be established.
+    // Default is 300 seconds.
+    curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 300);
+    auto res = curl_easy_perform(curl);
+    curl_easy_cleanup(curl);
+    fclose(fp);
+  
+    if (res != CURLE_OK) {
+      // TODO: Remove the temp file, then throw PXP-error here. 
+      fs::remove(path);
+      return;
+    }
+}
+
+// This method does the following. If the file matching the "filename" field of the
+// file_obj JSON does not exist OR if its hash does not match the sha value in the
+// "sha256" field of file_obj, then:
+//    (1) The file is downloaded using libcurl from the first entry in master-uris
+//        to a temporary directory. If this download fails, a PXP error is thrown.
+//
+//    (2) If the downloaded file's sha does not match the provided sha, then a PXP
+//        error is returned.
+//
+//    (3) If (1) and (2) both succeed, then the downloaded file is atomically
+//        renamed to tasks_cache_dir/<filename>
+//
+// The method returns the absolute path of the file. <-- The return value can be
+// removed, I put it there for convenience w/ MAIN_IMPL
+//
+// TODO: There is an inconsistency in orchestrator and the spec. of PCP-769. Orchestrator
+// puts the path under the "name" key of file_obj, while PCP-769 indicates that this is
+// "filename"
+/*static*/ string update_task_file(const fs::path& tasks_cache_dir, const lth_jc::JsonContainer& file_obj) {
+    auto filename = file_obj.get<string>("filename");
+    auto sha256 = file_obj.get<string>("sha256");
+    auto filepath = tasks_cache_dir / filename;
+    // TODO: Also check permissions? Make sure it's executable and readable/writable by
+    // PXP-agent's owner/group
+    if (fs::exists(filepath) && sha256 == calculate_sha256(filepath.string())) {
+      return filepath.string();
+    }
+  
+    // TODO: Not sure how I'd use code-id. 
```

Comment:
You shouldn't need to use code-id at all. You should append all entries in "params" as query params on the URL. You'll have to URI escape them as well. https://curl.haxx.se/libcurl/c/curl_easy_escape.html